### ISSUE TITLE
RGD decomposes two sidechains to single rgroup

### ISSUE THREAD

**Issue by jones-gareth** [rdkit/rdkit#5573]
-----------------------------------------

**Describe the bug**

In some cases RGD will decompose two sidechains attached to the same core atom into a single rgroup.  

The expected behavior is that each sidechain should be a separate rgroup (see test  `testSDFGroupMultiCoreNoneShouldMatch` and others).

**To Reproduce**

```python
# %%

from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem.rdRGroupDecomposition import RGroupDecomposition, RGroupDecompositionParameters, \
    RGroupMatching, RGroupScore, RGroupLabels, RGroupCoreAlignment, RGroupLabelling
from rdkit.Chem.Draw import IPythonConsole

IPythonConsole.ipython_useSVG = True

from rdkit.Chem import PandasTools
from rdkit.Chem import Draw
from IPython.display import HTML

import pandas as pd

# %%

core_block = """
  Mrv2008 09172211422D          

  8  8  0  0  0  0            999 V2000
   -7.2098   -3.1928    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -7.9243   -3.6053    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -7.9243   -4.4304    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -7.2098   -4.8429    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -6.4954   -4.4304    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -6.4954   -3.6053    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -5.7809   -4.8429    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -5.0664   -4.4304    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  1  6  1  0  0  0  0
  5  7  1  0  0  0  0
  7  8  1  0  0  0  0
M  END

"""
core = Chem.MolFromMolBlock(core_block)
core
# %%

test_block = """MolHeader
  Mrv2008 09172211422D          

 10 10  0  0  0  0            999 V2000
   -2.6437    1.7625    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.3582    1.3500    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.3582    0.5250    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.6438    0.1125    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -1.9293    0.5250    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.9293    1.3500    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -4.1730    1.4791    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.5085    2.1612    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.2148    0.1125    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -0.5004    0.5250    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  6  1  1  0  0  0  0
  2  7  1  0  0  0  0
  2  8  1  0  0  0  0
  5  9  1  0  0  0  0
  9 10  1  0  0  0  0
M  END

"""
test = Chem.MolFromMolBlock(test_block)
test

# %%

options = RGroupDecompositionParameters()
options.matchingStrategy = RGroupMatching.GreedyChunks
options.matchingStrategy = RGroupMatching.Exhaustive
options.onlyMatchAtRGroups = False
options.scoreMethod = RGroupScore.FingerprintVariance

decomp = RGroupDecomposition([core], options)
result1 = decomp.Add(test)
result1
# %%

decomp.Process()
groups = decomp.GetRGroupsAsColumns()
groups

# %%
df = PandasTools.RGroupDecompositionToFrame(groups, [test], include_core=True)
PandasTools.ChangeMoleculeRendering(df)
HTML(df.to_html())
```

**Expected behavior**

Expect to see two methyl rgroups `C[*:1]` and `C[*:2]`.  Instead we see a single rgroup `C[*:1].C[*:1]`

**Screenshots**

![image](https://user-images.githubusercontent.com/24233741/190870073-062298ca-b321-406e-8594-db9ccfde29bc.png)

**Configuration (please complete the following information):**
 - RDKit version: master as of 9.13.2022
 - OS: [e.g. Ubuntu 20.04]: Ubuntu 20.04
 - Python version (if relevant):
 - Are you using conda?
 - If you are using conda, which channel did you install the rdkit from?
 - If you are not using conda: how did you install the RDKit?

**Additional context**
I'll try and submit a PR for a fix

**Comment by bp-kelley** [comment]
It does the same thing when splitting a ring.  I’m not a fan of this behavior but I kind of understand it.

I have some cleanup code for molzip that fixes this, but only after the fact.  I’ll push the PR so you can check it out.

**Comment by jones-gareth** [comment]
> It does the same thing when splitting a ring. I’m not a fan of this behavior but I kind of understand it.
> 
> I have some cleanup code for molzip that fixes this, but only after the fact. I’ll push the PR so you can check it out.

Thanks.  I've also seen the cyclic case.  The odd thing is that most of the time it doesn't do this (there are plenty of examples in the test suite), so this is a bug.

The original RGD did have this behavior which we abandoned.  I remember a long discussion with Greg where we decided on one R group for each valence position.  I'd like to get rid of this behavior- it makes clustering the rgroups and doing downstream analyses like FW difficult.

If there are people that really want to do the R grouping by core atom it would make more sense to do a post-processing for them.

Haven't had time to debug yet.

**Comment by jones-gareth** [comment]
OK so I was completely mistaken- this is not a bug, the current default behavior is that two sidechains attached to a single core atom *will* decompose into a single rgroup *unless* there is a single user-defined attachment point on the core atom.  Perhaps I should move this issue into discussions.

This seems most unsatisfactory.  It looks like you or Paulo have previously noticed this in RGroupDecomp.cpp (line 291)

```cpp
      // it messes up when there are multiple ?
            if (!at->hasProp(_rgroupInputDummy)) {
```

Unless anyone objects I'll try and change the behavior when a parameter is set.  I think that can be done by adding explicit attachment dummy atoms to all possible groups in the input core.

**Comment by ptosco** [comment]
@jones-gareth That comment is not mine - `git blame` shows it's from Greg (2017). You are correct on the above - if there is not user-defined label, the current behavior is that a single R-group is added with two attachment points bearing the same label.
As @bp-kelley mentioned, I also have some local code to work around this limitation which otherwise causes problems when using `molzip`.
I agree that improving the current behavior by adding an optional flag that allows extracting as many R groups as the actual substituents would be good, provided that
1. if the new flag is not specified we stick to the historic behavior to avoid breaking existing code (I guess a number of people have workarounds implemented such as Brian's and mine that would break otherwise)
2. the labelling for the extracted groups is stable (e.g., if you have a methyl and an ethyl at a certain position they are _always_ extracted with the same label numbers)

**Comment by bp-kelley** [comment]
I think the basic solution is simple, however it potentially adds a new core which might be confusing.

However, i would propose that molzip should be able to reassemble any decomposition coming out of the rgd so that would set the standard and make for easier testing.

**Comment by jones-gareth** [comment]
@ptosco All your suggestions make sense to me except:

> the labelling for the extracted groups is stable (e.g., if you have a methyl and an ethyl at a certain position they are always extracted with the same label numbers)

I had assumed that we would assign positions based on the current strategy (greedy, greedychunks or exhaustive) and scoring metric.  So (except for exhaustive) the label numbers would assigned based on how the sidechains fitted to the decomposition of previously added compounds.  Assuming that works , I think it meets your criteria.

@bp-kelley 
>  potentially adds a new core which might be confusing.

I don't think it will require another core, but that's an implementation detail.  We wouldn't want to not have the feature because the implementation is awkward.  

> molzip should be able to reassemble any decomposition coming out of the rgd

I had assumed that, so I was unhappy to see molzip failing on RGD output.

If either of you could share your molzip patches that would be great!

**Comment by ptosco** [comment]
> I had assumed that we would assign positions based on the current strategy (greedy, greedychunks or exhaustive) and scoring metric. So (except for exhaustive) the label numbers would assigned based on how the sidechains fitted to the decomposition of previously added compounds. Assuming that works , I think it meets your criteria.
Yes, sure, that's correct, I have amended my comment to avoid confusion when coming back to this page in the future.

**Comment by bp-kelley** [comment]
Avengers power too low

**Comment by bp-kelley** [comment]
I’m not worried about complexity, I’m worried about correctness in this case.
